### PR TITLE
Fixes #37678 - Add postgresql-evr as a migration

### PR DESCRIPTION
**Pull request by sjha4**
#### What are the changes introduced in this pull request?
Early draft that works for new setups.
Need to work out existing installations.
#### Considerations taken when implementing this change?
We want to drop the postgres_evr extension and move all that to live inside katello.

#### What are the testing steps for this pull request?
1. Check `sudo rpm -qa *evr*`
2. Remove the rpm: `sudo rpm -e evr_package_from_above`
3. Run `bundle exec rails katello:reset` to mimic fresh installs
4. Test the usual flows with applicability

To test existing systems just delete the evr package and make sure everything continues to work as normal.

### CODE REVIEW

**Review by ianballou — APPROVED** [review]
This is working well for me! Tested on Katello nightly. Tested by resetting the database with and without postgresql-evr installed. I also ran applicability just to make sure it still works.

**Comment by sjha4** [comment]
Added a commit to handle existing systems with evr extension enabled.
The migration drops the extension which also drops the following constructs:
```
katello=# DROP EXTENSION evr CASCADE;
NOTICE:  drop cascades to 6 other objects
DETAIL:  drop cascades to column evr of table katello_rpms
drop cascades to column evr of table katello_installed_packages
drop cascades to trigger evr_insert_trigger_katello_rpms on table katello_rpms
drop cascades to trigger evr_update_trigger_katello_rpms on table katello_rpms
drop cascades to trigger evr_insert_trigger_katello_installed_packages on table katello_installed_packages
drop cascades to trigger evr_update_trigger_katello_installed_packages on table katello_installed_packages
```
Note that dropping extension deletes column evr on katello_rpms and katello_installed_packages.

The migration then recreates these constructs and repopulates the data in the recreated evr fields on above 2 tables using the new katello evr_t type.

Note: Not sure if this needs to be a reversible migration or not. 

To test for existing systems.

1. On master branch, create a host with some applicable errata.
2. Change to this branch and run db:migrate
3. Make sure the applicable errata still shows up on the host after running the migration on this branch and applicability works as expected.

**Comment by ianballou** [comment]
I just double checked that the triggers still work as well:

```
irb(main):004:0> ::Katello::InstalledPackage.first.update(epoch: "99999999")
=> true
irb(main):005:0> ::Katello::InstalledPackage.first
=> 
#<Katello::InstalledPackage:0x00007fc4499e6c80
 id: 1,
 name: "subscription-manager-rhsm-certificates",
 nvra: "subscription-manager-rhsm-certificates-20220623-1.el9.noarch",
 nvrea: "subscription-manager-rhsm-certificates-20220623-1.el9.noarch",
 epoch: "99999999",
 version: "20220623",
 release: "1.el9",
 arch: "noarch",
 vendor: "Red Hat, Inc.",
 evr: "(99999999,\"{\"\"(20220623,)\"\"}\",\"{\"\"(1,)\"\",\"\"(0,el)\"\",\"\"(9,)\"\"}\")">
irb(main):006:0> ::Katello::InstalledPackage.first.update(release: "98.el99")
=> true
irb(main):007:0> ::Katello::InstalledPackage.first
=> 
#<Katello::InstalledPackage:0x00007fc449507a98
 id: 1,
 name: "subscription-manager-rhsm-certificates",
 nvra: "subscription-manager-rhsm-certificates-20220623-1.el9.noarch",
 nvrea: "subscription-manager-rhsm-certificates-20220623-1.el9.noarch",
 epoch: "99999999",
 version: "20220623",
 release: "98.el99",
 arch: "noarch",
 vendor: "Red Hat, Inc.",
 evr: "(99999999,\"{\"\"(20220623,)\"\"}\",\"{\"\"(98,)\"\",\"\"(0,el)\"\",\"\"(99,)\"\"}\")">
```

**Comment by ekohl** [comment]
I see a failure related to this but not 100% sure if it's our infra or a bug: https://community.theforeman.org/t/katello-nightly-rpm-pipeline-2086-failed/38990/2

File: db/migrate/20240624121212_katello_recreate_evr_constructs.rb [hunk 1]
```diff
@@ -0,0 +1,159 @@
+class KatelloRecreateEvrConstructs < ActiveRecord::Migration[6.1]
+  def change
```
Comment (ekohl) [inline db/migrate/20240624121212_katello_recreate_evr_constructs.rb:2]: Should this have been `up` and then make it irreversible?

File: db/migrate/20240624121212_katello_recreate_evr_constructs.rb [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    count = select_value <<~SQL
+      SELECT count(*) FROM pg_extension WHERE extname = 'evr';
+    SQL
+    if count.to_i == 0
+      return
+    else
+      execute <<~SQL
+        DROP EXTENSION evr CASCADE;
+      SQL
```
Comment (ekohl) [inline db/migrate/20240624121212_katello_recreate_evr_constructs.rb:11]: There's a native DSL for this:
```suggestion
     disable_extension('evr', force: true)
```

That also properly reloads the type map.

File: db/migrate/20240624121212_katello_recreate_evr_constructs.rb [hunk 3]
```diff
@@ -0,0 +1,159 @@
+class KatelloRecreateEvrConstructs < ActiveRecord::Migration[6.1]
+  def change
+    count = select_value <<~SQL
+      SELECT count(*) FROM pg_extension WHERE extname = 'evr';
+    SQL
+    if count.to_i == 0
```
Comment (ekohl) [inline db/migrate/20240624121212_katello_recreate_evr_constructs.rb:6]: In the native DSL there's both `extension_available?` and `extension_enabled?`. I think the latter was intended here:
```suggestion
    if !extension_enabled?('evr')
```

**Review by ekohl — COMMENTED** [review]
I started to look at these specific lines because they failed. I'll do some further testing, but I think it makes sense to fix up the migration.

File: db/migrate/20240624121212_katello_recreate_evr_constructs.rb (hunk 2 above)
Comment (sjha4) [reply db/migrate/20240624121212_katello_recreate_evr_constructs.rb:11]: I will create a follow up PR with these updates..The earlier migration was indeed using enable_extension. 

Do you know if the DSL surpasses the privilege restrictions somehow?

Comment (ekohl) [reply db/migrate/20240624121212_katello_recreate_evr_constructs.rb:11]: No, it's using the same SQL

**Pull request merged**